Professora: Tatiana S. Tavares - Pós Tecnologias Disruptivas 2019/2

Módulo: Aprendizado Supervisionado, Machine Learning e Classificadores

Entrada: conjunto de dados MNIST. Entradas imagens 28 x 28

60000 para treinamento e 10000 para validação

MLPClassifier: Dense(512, activation='relu')

              Dense(10, activation='softmax')
metrics=['accuracy'] no treinamento e no test

In [1]:
from keras.datasets import mnist
(train_images, train_labels), (test_images, test_labels) = mnist.load_data()
print (train_images.shape)
print (test_images.shape)
print (train_labels)
print (test_labels)

Using TensorFlow backend.


11493376/11490434 [==============================] - 0s 0us/step
(60000, 28, 28)
(10000, 28, 28)
[5 0 4 ... 5 6 8]
[7 2 1 ... 4 5 6]


O fluxo de trabalho será o seguinte: primeiro, vamos alimentar a rede neural com os dados de treinamento, train_images e 
train_labels. A rede aprenderá então a associar imagens e rótulos. Por fim, pedimos que a rede produza previsões para as 
imagens de teste e verificamos se essas previsões correspondem aos rótulos de test_labels.

In [2]:
from keras import models
from keras import layers
network = models.Sequential()
network.add(layers.Dense(512, activation='relu', input_shape=(28 * 28,)))
network.add(layers.Dense(10, activation='softmax'))

Aqui, nossa rede consiste em uma sequência de duas camadas densas, que são camadas neurais densamente conectadas 
(também chamadas totalmente conectadas). A segunda (e última) camada é uma camada softmax de 10 vias, o que significa 
que ela retornará uma matriz de 10 pontuações de probabilidade (somando 1). Cada pontuação será a probabilidade de a 
imagem do dígito atual pertencer a uma das nossas classes de 10 dígitos.

Para tornar a rede pronta para o treinamento, precisamos escolher mais três coisas, como parte da etapa de compilação:
- Uma função de perda - como a rede poderá medir seu desempenho nos dados de treinamento e, assim, como será capaz de 
se orientar na direção certa.
- Um otimizador - O mecanismo pelo qual a rede se atualizará com base nos dados que vê e na sua função de perda.
- Métricas a serem monitoradas durante o treinamento e o teste: aqui, nos preocuparemos apenas com a acurácia (a fração 
das imagens que foram classificadas corretamente).

In [3]:
network.compile(optimizer='rmsprop', loss='categorical_crossentropy', metrics=['accuracy'])

Antes do treinamento, pré-processaremos os dados reformulando-os de acordo com o formato esperado pela rede e 
dimensionando-os para que todos os valores fiquem no intervalo [0, 1]. Anteriormente, nossas imagens de treinamento, 
por exemplo, eram armazenadas em uma matriz de forma (60000, 28, 28) do tipo uint8 com valores no intervalo [0, 255]. 
Nós o transformamos em uma matriz float32 (60000, 28 * 28) com valores entre 0 e 1.

In [0]:
train_images = train_images.reshape((60000, 28 * 28))
train_images = train_images.astype('float32') / 255
test_images = test_images.reshape((10000, 28 * 28))
test_images = test_images.astype('float32') / 255

In [0]:
print (train_images.shape)
print (test_images.shape)
print (train_labels)
print (test_labels)

Também precisamos codificar categoricamente os rótulos

In [5]:
from keras.utils import to_categorical
train_labels = to_categorical(train_labels)
test_labels = to_categorical(test_labels)

print (train_labels)
print (test_labels)

[[0. 0. 0. ... 0. 0. 0.]
 [1. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]
 ...
 [0. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 1. 0.]]
[[0. 0. 0. ... 1. 0. 0.]
 [0. 0. 1. ... 0. 0. 0.]
 [0. 1. 0. ... 0. 0. 0.]
 ...
 [0. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]]


Agora estamos prontos para treinar a rede, o que em Keras é feito por meio de uma chamada para o método 
de ajuste da rede. Nós ajustamos o modelo aos dados de treinamento:

In [6]:
network.fit(train_images, train_labels, epochs=100, batch_size=128)

Instructions for updating:
Use tf.where in 2.0, which has the same broadcast rule as np.where



Epoch 1/100





60000/60000 [==============================] - 8s 141us/step - loss: 0.2555 - acc: 0.9262
Epoch 2/100
60000/60000 [==============================] - 3s 47us/step - loss: 0.1041 - acc: 0.9692
Epoch 3/100
60000/60000 [==============================] - 3s 49us/step - loss: 0.0692 - acc: 0.9790
Epoch 4/100
60000/60000 [==============================] - 3s 46us/step - loss: 0.0498 - acc: 0.9846
Epoch 5/100
60000/60000 [==============================] - 3s 46us/step - loss: 0.0382 - acc: 0.9887
Epoch 6/100
60000/60000 [==============================] - 3s 44us/step - loss: 0.0292 - acc: 0.9916
Epoch 7/100
60000/60000 [==============================] - 3s 46us/step - loss: 0.0221 - acc: 0.9932
Epoch 8/100
60000/60000 [==============================] - 3s 48us/step - loss: 0.0170 - acc: 0.9953
Epoch 9/100
60000/60000 [==============================] - 3s 44us/step - loss: 0.0137 - 

In [7]:
test_loss, test_acc = network.evaluate(test_images, test_labels)
print('test_acc:', test_acc)

10000/10000 [==============================] - 1s 64us/step
test_acc: 0.9847


A acurácia do conjunto de testes é de aproximadamente 98% - abaixo da precisão do conjunto de treinamento (~99%). 
Essa lacuna entre a precisão do treinamento e a precisão do teste é um exemplo de
overfitting: o fato de que os modelos de aprendizado de máquina tendem a ter um desempenho pior em novos dados 
do que em seus dados de treinamento.

